src/mlpack/methods/ann/activation_functions/gcu_function.hpp

```diff
@@ -0,0 +1,79 @@
+/**
+ * @file methods/ann/activation_functions/gcu_function.hpp
+ * @author Tarek Elsayed
+ *
+ * Definition and implementation of the Growing Cosine Unit function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_GCU_FUNCTION_HPP
+#define MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_GCU_FUNCTION_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+
+/**
+ * The gcu function, defined by
+ *
+ * @f{eqnarray*}{
+ * f(x) &=& x * cos(x) \\
+ * f'(x) &=& cos(x) - x sin(x) \\
+ * @f}
+ */
+class GCUFunction
+{
+ public:
+  /**
+   * Computes the gcu function.
+   *
+   * @param x Input data.
+   * @return f(x).
+   */
+  static double Fn(const double x)
+  {
+    return x * std::cos(x);
+  }
+
+  /**
+   * Computes the gcu function.
+   *
+   * @param x Input data.
+   * @param y The resulting output activation.
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Fn(const InputVecType& x, OutputVecType& y)
+  {
+    y = x % arma::cos(x);
+  }

```

Comment:
would you explain what is the difference between these two?

Reply:
In the first one, there is a multiply operation, while in the second it is a reminder operator that is used. Why?

Reply:
% is the operator for element-wise multiplication of two objects in Armadillo.

Reply:
> would you explain what is the difference between these two?

First function takes a single value (double) as input. Second one takes a matrix as input.